tc/autotuner/autotuner-inl.h

```diff
@@ -0,0 +1,560 @@
+        auto config = detail::makeTuningConfiguration<Backend>(
+            options, modelConfiguration);
+        config.fixParameters(fixedParams);
+        return config;
+      });
+
+  // searchStrategy is passed to tuningHarness.run()
+  SearchStrategy searchStrategy(
+      configs,
+      FLAGS_tuner_gen_generations,
+      FLAGS_tuner_gen_pop_size,
+      FLAGS_tuner_gen_crossover_rate,
+      FLAGS_tuner_gen_mutation_rate,
+      FLAGS_tuner_gen_number_elites);
+
+  // Create a tuning harness
+  detail::TuningHarness<Backend> tuningHarness(
+      FLAGS_tuner_gen_pop_size,
+      tcEntryPointMap_.at(tcEntryPoint),
+      inputs,
+      outputs,
+      baseMapping,
+      fixedParams,
+      optionsCache_);
+
+  // Setup handlers
+  sigterm_ = 0;
+  sigint_ = 0;
+  auto sigtermOrigHandler = std::signal(SIGTERM, [](int) { sigterm_ = 1; });
+  auto sigintOrigHandler = std::signal(SIGINT, [](int) { sigint_ = 1; });
+  ScopeGuard handlersGuard([&]() {
+    std::signal(SIGTERM, sigtermOrigHandler);
+    std::signal(SIGINT, sigintOrigHandler);
+  });
+
+  // Run harness in a separate thread
+  std::atomic_bool tuningHarnessFinished(false);
+  std::exception_ptr tuningHarnessThreadEx = nullptr;
+  std::thread tuningHarnessThread([&]() {
+    try {
+      tuningHarness.run(searchStrategy);
+    } catch (const std::exception& e) {
+      std::cerr << "Exception during autotuning: " << e.what()
+                << "\n dumping cache to "
+                << tc::makeOptionsFilename(cacheFileName) << "/"
+                << Backend::makeDeviceFilename(cacheFileName) << std::endl;
+      storeCaches<Backend>(optionsCache_, cacheFileName);
+      tuningHarnessThreadEx = std::current_exception();
+    }
+    tuningHarnessFinished = true;
+  });
+  while (not tuningHarnessFinished) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    if (sigint_) {
+      std::cerr
+          << "Autotuning will stop after the current iteration has finished."
+          << std::endl;
+      tuningHarness.stopAfterCurrentIteration();
+      tuningHarnessThread.join();
+      storeCaches<Backend>(optionsCache_, cacheFileName);
+    }
+    if (sigterm_) {
+      std::cerr << "Autotuning aborted." << std::endl;
+      storeCaches<Backend>(optionsCache_, cacheFileName);
+      std::abort();
+    }
+  }
+
+  tuningHarnessThread.join();
+
+  if (tuningHarnessThreadEx) {
+    std::rethrow_exception(tuningHarnessThreadEx);
+  }
+
+  // only store cache if the file path is provided
+  if (!cacheFileName.empty()) {
+    storeCaches<Backend>(optionsCache_, cacheFileName);
+  }
+
+  CHECK_GT(inputs.size(), 0);
+  auto inputInfos = makeTensorInfoVector(inputs.begin()->second);
+  auto outputInfos = makeTensorInfoVector(outputs.begin()->second);
+  return optionsCache_->getTopKOptions(
+      canonicalTc(tcEntryPointMap_.at(tcEntryPoint)),
+      inputInfos,
+      outputInfos,
+      Backend::deviceString(),
+      1);
```

Comment:
why doesn't this use `harness.bestMappingOptions()` ?  It looks like we have two ways of doing the same thing.